onnxruntime/core/framework/session_state_initializer.cc

```diff
@@ -181,37 +184,46 @@ common::Status SaveMLValueNameIndexMapping(const GraphViewer& graph_viewer,
   if (strcmp(alloc_info.name, CPU) == 0 || alloc_info.mem_type == OrtMemTypeCPUOutput) {
     // deserialize directly to CPU tensor
-    return utils::TensorProtoToMLValue(tensor_proto, alloc_ptr, preallocated, preallocated_size, mlvalue);
+    return utils::TensorProtoToMLValue(env, proto_path.c_str(), tensor_proto, m, mlvalue, deleter);
   }
-
-  std::unique_ptr<Tensor> p_tensor;
-  // deserialize to CPU first for non-CPU allocator, then alloc and copy
-  AllocatorPtr deserialize_alloc_ptr;
-  std::unique_ptr<Tensor> p_deserialize_tensor;
-  deserialize_alloc_ptr = exec_providers.Get(kCpuExecutionProvider)->GetAllocator(0, OrtMemTypeDefault);
-  ORT_RETURN_IF_ERROR(utils::GetTensorFromTensorProto(tensor_proto, &p_deserialize_tensor,
-                                                      deserialize_alloc_ptr));
+  //alloc_info.name is not 'CPU'
   const IExecutionProvider* provider = exec_providers.Get(alloc_info);
-  ORT_ENFORCE(provider != nullptr);
-  p_tensor = std::make_unique<Tensor>(
-      p_deserialize_tensor->DataType(),
-      p_deserialize_tensor->Shape(),
-      preallocated ? preallocated : static_cast<void*>(alloc_ptr->Alloc(p_deserialize_tensor->Size())),
-      alloc_info,
-      preallocated ? nullptr : alloc_ptr);  // no deleter for preallocated
-
-  Status copy_status = provider->CopyTensor(*p_deserialize_tensor, *p_tensor);
+  if (provider == nullptr) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Invalid allocation info. Provider name = ", alloc_info.name);
+  }
+  // deserialize and copy. In the copy stage, it won't check if the buffer has enough room.
+  // The result tensor won't need a deleter because:
+  // 1. It mustn't be a string tensor
+  // 2. The memory is not memory-mapped.
+  deleter.f = nullptr;
+  deleter.param = nullptr;
+  if (tensor_proto.data_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "string tensor is not supported for copying between allocators");
+  }
+  // deserialize to CPU first for non-CPU allocator, then alloc and copy
+  size_t cpu_tensor_length;
+  ORT_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<0>(tensor_proto, &cpu_tensor_length));
+  OrtAllocatorInfo info("Cpu", OrtDeviceAllocator, 0, OrtMemTypeDefault);
+  std::unique_ptr<char[]> data(new char[cpu_tensor_length]);
+  std::unique_ptr<Tensor> p_tensor;
+  MLValue tmp_mlvalue;
+  OrtCallback d;
+  ORT_RETURN_IF_ERROR(utils::TensorProtoToMLValue(
+      env, proto_path.c_str(), tensor_proto, MemBuffer(data.get(), cpu_tensor_length, info), tmp_mlvalue, d));
+  const Tensor& p_deserialize_tensor = tmp_mlvalue.Get<Tensor>();
+
+  p_tensor = std::make_unique<Tensor>(p_deserialize_tensor.DataType(), p_deserialize_tensor.Shape(), m.GetBuffer(),
+                                      m.GetAllocInfo());
+  // TODO: does this function work for string tensor?
+  Status copy_status = provider->CopyTensor(p_deserialize_tensor, *p_tensor);
```

Comment:
Does it still need validation of whether it works with a string?

Reply:
Yes. I don't know if it works. It depends on how the provider implements this function.